### PR TITLE
Add Support for ED25519 signature verification

### PR DESCRIPTION
**Pull request by symorton**
**Problem:**
We need to support ED25519 signature verification along with Secp256k1. The way the crypto layer was implemented, it was not extensible enough to allow multiple algorithm support in a scalable and simple way.

**Solution:**
- Changed the Secp256k1 Token Verifier to a more generic `TokenVerifier`. Introduced some abstractions through the CryptoOperations class to allow for multiple algorithms for signature verification. We will continue this work for Signers in subsequent PRs.
- Separated CryptoOperations into two classes: CryptoOperations for signature verification and other crypto operations, and KeyManagementOperations for all methods used for key management. 
- The `Jwk` class was introduced during backup and restore but was incompatible with other operations related to JSON web keys. This class has been renamed to `JWK` (following coding convention of sister Android lib), and integrated into more places where a JSON web key data model is used. More clean up to use JWK instead of `ECPublicKey` will be added in subsequent PRs. 

**Validation:**
All unit tests pass. 
Scenario for both Secp256k1 signatures and ED25519 signature were tested and work E2E

**Type of change:**
- [X] Feature work
- [ ] Bug fix
- [ ] Documentation
- [X] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

**Risk**:
- [X] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

**Work Item links**:
Please include here links for this work item, or deferred work, or related work. E.g. if the refactoring is too big to fit in this PR, or the localized strings need to be updated later, please link the TODO work items here.

**Documentation Links**:
Please include here links to any related background documentation for this PR.

### CODE REVIEW

File: VCCrypto/VCCrypto/Algo/Secp256k1.swift [hunk 1]
```diff
@@ -72,13 +90,14 @@ public struct Secp256k1: Signing {
...
     /// - Returns: True if the signature is valid
-    public func isValidSignature(signature: Data, forMessageHash messageHash: Data, usingPublicKey publicKey: Secp256k1PublicKey) throws -> Bool {
+    public func isValidSignature(signature: Data, forMessage message: Data) throws -> Bool {
         // Validate params
         guard signature.count == 64 else { throw Secp256k1Error.invalidSignature }
-        guard messageHash.count == 32 else { throw Secp256k1Error.invalidMessageHash }
-        
+        guard message.count == 32 else { throw Secp256k1Error.invalidMessageHash }
+    
+        let publicKey = self.publicKey == nil ? try createPublicKey() : self.publicKey!
```
Comment (dangodb) [inline VCCrypto/VCCrypto/Algo/Secp256k1.swift:99]: Should this be done in the constructor?

File: VCCrypto/VCCrypto/Algo/Secp256k1.swift [hunk 2]
```diff
@@ -18,19 +18,37 @@ enum Secp256k1Error: Error {
...
     
-    public init() {}
+    private let publicKey: Secp256k1PublicKey?
+    
+    private let secret: VCCryptoSecret?
+    
+    private let hashOperation = Sha256()
+    
+    /// Create Secp256k1 object from Secret, for signing.
+    public init(secret: VCCryptoSecret) throws {
```
Comment (dangodb) [inline VCCrypto/VCCrypto/Algo/Secp256k1.swift:28]: Curious, why pull these to the constructor?

File: VCCrypto/VCCrypto/CryptoOperations.swift [hunk 3]
```diff
@@ -3,78 +3,56 @@
...
-        let accessGroup = sdkConfiguration.accessGroupIdentifier
-        return Random32BytesSecret(withStore: secretStore, andId: id, inAccessGroup: accessGroup)
+    public func verify(signature: Data,
+                       forMessage message: Data,
+                       usingPublicKey publicKey: PublicKey) throws -> Bool {
+        
+        let algorithm = try getAlgorithm(publicKey: publicKey)
+        
+        /// If algorithm is secp256k1, hash message value.
+        var msg = message
```
Comment (dangodb) [inline VCCrypto/VCCrypto/CryptoOperations.swift:42]: Should this be done inside the secp256k class?

File: VCEntities/VCEntities/backup/content/microsoft2020/RawIdentity.swift [hunk 4]
```diff
@@ -65,18 +65,18 @@ struct RawIdentity: Codable {
                           alias: self.name)
     }
     
-    func jwkFromKeyContainer(_ keyContainer: KeyContainer) throws -> Jwk {
+    func jwkFromKeyContainer(_ keyContainer: KeyContainer) throws -> JWK {
 
         // Get out the private and public components of the key (pair)
         let secret = keyContainer.keyReference
-        let (privateKey, publicKey) = try Secp256k1().createKeyPair(forSecret: secret)
+        let (privateKey, publicKey) = try Secp256k1(secret: secret).createKeyPair(forSecret: secret)
```
Comment (dangodb) [inline VCEntities/VCEntities/backup/content/microsoft2020/RawIdentity.swift:72]: Why do we need to pass the secret in both the constructor and method?

File: VCToken/VCToken/JwsToken.swift [hunk 5]
```diff
@@ -65,10 +65,15 @@ public struct JwsToken<T: Claims> {
         self.signature = try signer.sign(token: self, withSecret: secret)
     }
     
-    public func verify(using verifier: TokenVerifying, withPublicKey key: ECPublicJwk) throws -> Bool {
+    public func verify(using verifier: TokenVerifying, withPublicKey key: JWK) throws -> Bool {
         return try verifier.verify(token: self, usingPublicKey: key)
     }
     
+    /// Temporary: TODO: remove support for ECPublicJwk data model for JWK.
```
Comment (dangodb) [inline VCToken/VCToken/JwsToken.swift:72]: Expected? Do we have a bug tracking this?

File: VCToken/VCToken/algorithms/ECPublicJwk.swift [hunk 6]
```diff
@@ -41,8 +41,30 @@ public struct ECPublicJwk: Codable, Equatable {
...
+                   x: Data(base64URLEncoded: x),
+                   y: encodedY)
+    }
+    
     func getMinimumAlphabeticJwk() -> String {
-        return "{\"crv\":\"\(self.curve)\",\"kty\":\"\(self.keyType)\",\"x\":\"\(self.x)\",\"y\":\"\(self.y)\"}"
+        
+        var encodedJwk = "{\"crv\":\"\(curve)\",\"kty\":\"\(keyType)\",\"x\":\"\(x)\""
+        
+        if let y = y {
```
Comment (dangodb) [inline VCToken/VCToken/algorithms/ECPublicJwk.swift:63]: What is this case?

File: VCCrypto/VCCrypto/CryptoOperations.swift [hunk 7]
```diff
@@ -3,78 +3,56 @@
...
+    case invalidPublicKey
 }
 
+/// Operations that are involved in verification cryptographic operations..
 public struct CryptoOperations: CryptoOperating {
-
-    private let secretStore: SecretStoring
     
-    private let sdkConfiguration: VCSDKConfigurable
+    public init() {}
```
Comment (dangodb) [inline VCCrypto/VCCrypto/CryptoOperations.swift:13]: Do we need unit tests for this class? And the Key management one?

File: VCToken/VCToken/algorithms/TokenVerifier.swift [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+*--------------------------------------------------------------------------------------------*/
+
+import VCCrypto
+
+enum TokenVerifierError: Error {
+    case unsupportedAlgorithmFoundInJWK
+    case malformedJWK
+}
+
+public struct TokenVerifier: TokenVerifying {
```
Comment (dangodb) [inline VCToken/VCToken/algorithms/TokenVerifier.swift:13]: Do we need unit tests for this class?

**Review by dangodb — CHANGES_REQUESTED** [review]

File: VCCrypto/VCCrypto/Algo/Secp256k1.swift (hunk 1 above)
Comment (symorton) [reply VCCrypto/VCCrypto/Algo/Secp256k1.swift]: reverted back to old implementation

File: VCCrypto/VCCrypto/Algo/Secp256k1.swift (hunk 2 above)
Comment (symorton) [reply VCCrypto/VCCrypto/Algo/Secp256k1.swift]: reverted back to old implementation per discussion

File: VCCrypto/VCCrypto/CryptoOperations.swift (hunk 3 above)
Comment (symorton) [reply VCCrypto/VCCrypto/CryptoOperations.swift]: per discussion created a ES256k class that does hashing and secp256k1 signing

File: VCEntities/VCEntities/backup/content/microsoft2020/RawIdentity.swift (hunk 4 above)
Comment (symorton) [reply VCEntities/VCEntities/backup/content/microsoft2020/RawIdentity.swift]: reverted.

File: VCToken/VCToken/algorithms/ECPublicJwk.swift (hunk 6 above)
Comment (symorton) [reply VCToken/VCToken/algorithms/ECPublicJwk.swift:64]: added comment

File: VCCrypto/VCCrypto/CryptoOperations.swift (hunk 7 above)
Comment (symorton) [reply VCCrypto/VCCrypto/CryptoOperations.swift]: tests added.

File: VCToken/VCToken/algorithms/TokenVerifier.swift (hunk 8 above)
Comment (symorton) [reply VCToken/VCToken/algorithms/TokenVerifier.swift]: added

File: VCToken/VCToken/JwsToken.swift (hunk 5 above)
Comment (symorton) [reply VCToken/VCToken/JwsToken.swift:72]: bug added.

**Review by dangodb — APPROVED** [review]

**Pull request merged**